pkg/workload/debug/webhook_server_slow.go

```diff
@@ -0,0 +1,180 @@
+
+	const (
+		defaultTransientErrorFrequency = 1000 // Default transient error frequency in ms
+	)
+	defaultRangeDelays := []time.Duration{} // Default range delays in ms (no delay)
+
+	// Parse the arguments
+	transientErrorFrequency := defaultTransientErrorFrequency
+	rangeDelays := defaultRangeDelays
+
+	if len(args) > 0 {
+		if freq, err := strconv.Atoi(args[0]); err == nil {
+			transientErrorFrequency = freq
+		} else {
+			log.Printf("Invalid transient error frequency, using default: %d ms", defaultTransientErrorFrequency)
+		}
+	}
+
+	if len(args) > 1 {
+		rangeDelays = make([]time.Duration, len(args[1:]))
+		for i, delay := range args[1:] {
+			if d, err := strconv.Atoi(delay); err == nil {
+				rangeDelays[i] = time.Duration(d) * time.Millisecond
+			} else {
+				log.Printf("Invalid range delay at index %d, using default: 0 ms", i)
+				rangeDelays[i] = 0 * time.Millisecond
+			}
+		}
+	}
+
+	// Log the parsed or default values
+	log.Printf("Transient Error Frequency: %d ms", transientErrorFrequency)
+	log.Printf("Range Delays: %v ms", rangeDelays)
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		var req struct {
+			Length  int `json:"length"`
+			Payload []struct {
+				After struct {
+					ID  int `json:"id"`
+					VAL int `json:"val"`
+				} `json:"after"`
+				Before struct {
+					ID  int `json:"id"`
+					VAL int `json:"val"`
+				} `json:"before"`
+				Updated string `json:"updated"`
+			} `json:"payload"`
+		}
+
+		err := json.NewDecoder(r.Body).Decode(&req)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		now := timeutil.Now()
+		if now.Sub(lastTransientErrorTime) >= time.Duration(transientErrorFrequency)*time.Millisecond {
+			lastTransientErrorTime = now
+			http.Error(w, "transient sink error", http.StatusInternalServerError)
+			return
+		}
+
+		var before, after, d int
+		func() {
+			mu.Lock()
+			defer mu.Unlock()
+			before = len(seen)
+			after = before
+			// TODO(cdc): add check for ordering guarantees using resolved timestamps and event timestamps
+			for _, i := range req.Payload {
+				id := i.After.ID
+				seenKey := fmt.Sprintf("%d-%s", id, i.Updated)
+				if _, ok := seen[seenKey]; !ok {
+					seen[seenKey] = struct{}{}
+					after++
+
+					sleepDurationIndex := id / 10
+					if sleepDurationIndex < len(rangeDelays) {
+						timeToSleep := rangeDelays[sleepDurationIndex]
+						time.Sleep(timeToSleep)
+					}
+				} else {
+					dupes++
+				}
+			}
+			if r.ContentLength > 0 {
+				size += r.ContentLength
+			}
+			d = dupes
+		}()
+		const printEvery = 100
+		if before/printEvery != after/printEvery {
+			log.Printf("keys seen: %d (%d dupes); %.1f MB", after, d, float64(size)/float64(1<<20))
+		}
+	})
+	mux.HandleFunc("/reset", func(w http.ResponseWriter, r *http.Request) {
+		func() {
+			mu.Lock()
+			defer mu.Unlock()
+			seen = make(map[string]struct{}, len(seen))
+			dupes = 0
+			size = 0
```

Comment:
nit: having this mutex protecting these variables is a bit odd. can you move them all into a state struct which is mutex protected, just for extra clarity